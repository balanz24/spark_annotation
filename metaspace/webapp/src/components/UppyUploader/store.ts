// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
