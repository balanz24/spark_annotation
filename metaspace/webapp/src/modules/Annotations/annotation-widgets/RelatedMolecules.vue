<template>
  <div
    v-loading="loading"
    class="related-molecules"
  >
    <div
      v-for="other in sortedAnnotations"
      :key="other.ion"
    >
      <el-divider v-if="sortedAnnotations.length > 1">
        <div class="ion-heading">
          <component
            :is="other.ion !== annotation.ion ? 'router-link' : 'div'"
            :to="other.ion !== annotation.ion ? linkToAnnotation(other) : undefined"
            class="ion-link"
          >
            <div>
              <span v-if="other.isIsomer">Isomer:</span>
              <span v-else-if="other.isIsobar">
                Isobar
                <el-popover trigger="hover">
                  <span slot="reference">{{ renderMassShift(annotation.mz, other.mz) }}</span>
                  <p><molecular-formula :ion="annotation.ion" /> (Selected annotation): {{ annotation.mz.toFixed(4) }}</p>
                  <p><molecular-formula :ion="other.ion" /> (Isobar): {{ other.mz.toFixed(4) }}</p>
                  <p>Check the <b>Diagnostics</b> panel for more detail about the overlapping peaks.</p>
                </el-popover>:
              </span>
              <molecular-formula
                class="ion-formula"
                :ion="other.ion"
              />
            </div>

            <fdr-badge
              v-if="other.fdrLevel !== null && !hideFdr"
              :fdr-level="other.fdrLevel"
            />
            <msm-badge
              v-if="other.isIsobar"
              :msm-score="other.msmScore"
            />
          </component>

          <el-popover
            v-if="other.isIsomer"
            trigger="hover"
            placement="top"
          >
            <div style="max-width: 500px;">
              <p>
                The False Discovery Rate (FDR) for each annotation is calculated among all ions that share the same adduct.
              </p>
              <p>
                It is possible for isomeric annotations to show different FDRs due to having different adducts.
                In these cases, it is an indicator of certain adducts having a higher or lower probability of
                incorrectly labelling unknown molecules.
              </p>
              <p>
                The FDR should not be used to decide which isomeric molecule is more likely to be correct.
              </p>
            </div>
            <i
              slot="reference"
              class="el-icon-question help-icon"
            />
          </el-popover>

          <el-popover
            v-else-if="other.isIsobar"
            trigger="hover"
            placement="top"
          >
            <div style="max-width: 400px;">
              <p>
                When two isobaric ions are annotated with significantly different MSM scores (>0.5),
                it is generally reasonable to assume that the lower-scoring ion is a false discovery.
              </p>
              <p>
                To help manually review cases when the MSM scores are not significantly different,
                the <b>Diagnostics</b> panel allows side-by-side comparison of isotopic ion images and spectra.
              </p>
            </div>
            <i
              slot="reference"
              class="el-icon-question help-icon"
            />
          </el-popover>
        </div>
      </el-divider>

      <compounds-list :compounds="other.possibleCompounds" />
    </div>

    <p
      v-if="sortedAnnotations != null && sortedAnnotations.length === 0"
      class="empty-message"
    >
      No annotations found.
    </p>
  </div>
</template>

<script>
import { omit, sortBy, uniqBy } from 'lodash-es'
import { renderMassShift } from '../../../lib/util'
import { relatedMoleculesQuery } from '../../../api/annotation'
import { encodeParams, stripFilteringParams } from '../../Filters'
import { ANNOTATION_SPECIFIC_FILTERS } from '../../Filters/filterSpecs'
import CompoundsList from './CompoundsList.vue'
import FdrBadge from './FdrBadge.vue'
import MsmBadge from './MsmBadge.vue'
import config from '../../../lib/config'
import MolecularFormula from '../../../components/MolecularFormula'

export default {
  components: {
    CompoundsList,
    FdrBadge,
    MsmBadge,
    MolecularFormula,
  },
  props: {
    annotation: { type: Object, required: true },
    annotations: { type: Array, required: false },
    databaseId: { type: Number, required: true },
    hideFdr: { type: Boolean, default: false },
  },
  data() {
    return {
      loading: 0,
    }
  },
  apollo: {
    isomerAnnotations: {
      query: relatedMoleculesQuery,
      loadingKey: 'loading',
      skip() {
        return !config.features.isomers
      },
      variables() {
        return {
          datasetId: this.annotation.dataset.id,
          filter: { databaseId: this.databaseId, ionFormula: this.annotation.ionFormula },
          orderBy: 'ORDER_BY_FDR_MSM',
          sortingOrder: 'ASCENDING',
        }
      },
      update(data) {
        return data.allAnnotations
      },
    },
    isobarAnnotations: {
      query: relatedMoleculesQuery,
      loadingKey: 'loading',
      skip() {
        return !config.features.isobars
      },
      variables() {
        return {
          datasetId: this.annotation.dataset.id,
          filter: { databaseId: this.databaseId, isobaricWith: this.annotation.ionFormula },
          orderBy: 'ORDER_BY_FDR_MSM',
          sortingOrder: 'ASCENDING',
        }
      },
      update(data) {
        return data.allAnnotations
      },
    },
  },
  computed: {
    sortedAnnotations() {
      let annotations = this.annotations ? [
        ...this.annotations,
        ...(this.isomerAnnotations || []).map(ann => ({ ...ann, isIsomer: true })),
        ...(this.isobarAnnotations || []).map(ann => ({ ...ann, isIsobar: true })),
      ] : [
        this.annotation,
        ...(this.isomerAnnotations || []).map(ann => ({ ...ann, isIsomer: true })),
        ...(this.isobarAnnotations || []).map(ann => ({ ...ann, isIsobar: true })),
      ]
      annotations = uniqBy(annotations, a => a.ion)
      // Sort order: reference annotation first, then best by FDR, then best by MSM
      // (consistent with the default sort in AnnotationsTable and RelatedMolecules)
      annotations = sortBy(annotations,
        a => a.ion === this.annotation.ion ? 0 : 1,
        a => a.fdrLevel,
        a => -a.msmScore,
      )

      return annotations
    },
  },
  methods: {
    renderMassShift,
    linkToAnnotation(other) {
      const filters = {
        datasetIds: this.annotations ? this.annotations.map((annotation) => annotation.dataset.id)
          : [this.annotation.dataset.id],
        compoundName: other.sumFormula,
        chemMod: other.chemMod,
        neutralLoss: other.neutralLoss,
        adduct: other.adduct,
        fdrLevel: Math.max(other.fdrLevel, this.$store.getters.filter.fdrLevel || 0),
      }

      // Make a best effort to remove existing filters that might prevent showing the linked annotation, while
      // keeping the rest. e.g. keep expanded sections and selected FDR, but remove page and m/z
      const nonFilterParams = stripFilteringParams(this.$route.query)
      const filtersToKeep = omit(this.$store.getters.filter, ANNOTATION_SPECIFIC_FILTERS)
      const filterParams = encodeParams({
        ...filtersToKeep,
        ...filters,
      }, this.$route.path, this.$store.state.filterLists)
      return {
        query: {
          ...nonFilterParams,
          ...filterParams,
          page: undefined,
          sort: undefined,
        },
      }
    },
  },
}
</script>

<style scoped lang="scss">
  @import "~element-ui/packages/theme-chalk/src/common/var";

  .ion-heading {
    display: flex;
    align-items: center;
  }
  .ion-link, a.ion-link:link {
    display: flex;
    align-items: center;
    text-decoration: none;
    white-space: nowrap;
    font-size: 1.2em;
    color: inherit;

    .ion-formula {
      font-weight: bold;
    }
    .mass-shift {
      font-weight: bold;
    }
  }
  .help-icon {
    font-size: 16px;
    color: inherit;
  }
</style>
