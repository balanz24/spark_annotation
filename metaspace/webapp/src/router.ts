/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
