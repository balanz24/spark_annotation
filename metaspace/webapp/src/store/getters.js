import { mzFilterPrecision } from '../lib/util';
import { decodeParams, decodeSettings, getLevel } from '../modules/Filters';
import config from '../lib/config';

/** For filters where empty string is a valid client-side value that must be converted to empty string for the API */
const noneToEmptyString = s => s === 'none' ? '' : s;

export default {
  filterLevel(state) {
    return getLevel(state.route.path);
  },

  filter(state) {
    return decodeParams(state.route, state.filterLists);
  },

  currentUser(state) {
    return state.currentUser;
  },

  settings(state) {
    return decodeSettings(state.route);
  },

  ftsQuery(state, getters) {
    return getters.filter.simpleQuery;
  },

  gqlAnnotationFilter(state, getters) {
    const filter = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;

    const f = {
      compoundQuery: filter.compoundName,
      termId: filter.term ? parseInt(filter.term, 10) : undefined,
      chemMod: noneToEmptyString(filter.chemMod),
      neutralLoss: noneToEmptyString(filter.neutralLoss),
      adduct: filter.adduct,
      annotationId: filter.annotationIds ? filter.annotationIds.join('|') : undefined,
      fdrLevel: filter.fdrLevel,
      pValue: filter.pValue,
      colocalizedWith: filter.colocalizedWith,
      // Only include colocalizationAlgo if there is another filter that uses it. Otherwise the annotations list
      // refreshes unnecessarily when changing algorithm.
      colocalizationAlgo: filter.colocalizedWith || filter.colocalizationSamples ? colocalizationAlgo : null,
      colocalizationSamples: filter.colocalizationSamples,
      offSample: filter.offSample == null ? undefined : !!filter.offSample,
    };

    if (!config.features.neutral_losses && !config.features.advanced_ds_config) {
      f.hasNeutralLoss = false;
    }

    if (!config.features.chem_mods && !config.features.advanced_ds_config) {
      f.hasChemMod = false;
    }

    if (!config.features.all_adducts) {
      f.hasHiddenAdduct = false;
    }

    if (filter.minMSM)
      f.msmScoreFilter = { min: filter.minMSM, max: 1.0 };

    if (filter.mz) {
      const mz = parseFloat(filter.mz),
        deltamz = parseFloat(mzFilterPrecision(filter.mz));
      f.mzFilter = {
        min: (mz - deltamz),
        max: (mz + deltamz)
      };
    }

    f.databaseId = filter.database

    return f;
  },

  gqlDatasetFilter(state, getters) {
    const filter = getters.filter;
    const { group, project, submitter, datasetIds, polarity,
      organism, organismPart, condition, growthConditions,
      ionisationSource, analyzerType, maldiMatrix, metadataType,
      compoundName, datasetOwner, opticalImage } = filter;
    const level = getters.filterLevel;
    const isLogged = state.currentUser && state.currentUser.id
    const hasAnnotationMatching = level === 'dataset' && compoundName ? { compoundQuery: compoundName } : undefined;
    return {
      group: datasetOwner && datasetOwner !== 'my-datasets' && isLogged ? datasetOwner : group,
      project: project,
      submitter: datasetOwner === 'my-datasets' && isLogged ? state.currentUser.id : submitter,

      // temporary workaround because of array-related bugs in apollo-client
      ids: datasetIds ? datasetIds.join("|") : null,

      organism,
      opticalImage,
      organismPart,
      condition,
      growthConditions,
      ionisationSource,
      maldiMatrix,
      analyzerType,
      polarity: polarity ? polarity.toUpperCase() : null,
      metadataType,
      hasAnnotationMatching,
    }
  },

  gqlColocalizationFilter(state, getters) {
    const { datasetIds, colocalizedWith, database, fdrLevel } = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;
    if (datasetIds && !datasetIds.includes('|') && colocalizedWith != null && database != null && fdrLevel != null) {
      return { colocalizedWith, colocalizationAlgo, databaseId: database, fdrLevel };
    } else {
      return null;
    }
  },
}
