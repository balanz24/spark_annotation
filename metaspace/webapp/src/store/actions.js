import apolloClient from '../api/graphqlClient';
import {fetchOptionListsQuery} from '../api/metadata';
import {decodeParams} from '../modules/Filters';
import {computed} from "@vue/composition-api";
import store from "../store/index";
export default {

  async initFilterLists(context) {
    if(context.state.filterListsLoading || context.state.filterLists != null)
      return;

    context.commit('setFilterListsLoading');

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    });

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds)} );

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)
    context.commit('updateFilter', filter)
  },

};
