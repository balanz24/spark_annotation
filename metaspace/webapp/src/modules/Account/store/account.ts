import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
