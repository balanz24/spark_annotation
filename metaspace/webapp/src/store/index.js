import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
