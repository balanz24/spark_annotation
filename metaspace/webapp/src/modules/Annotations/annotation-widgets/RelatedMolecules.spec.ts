import { mount } from '@vue/test-utils'
import RelatedMolecules from './RelatedMolecules.vue'
import { initMockGraphqlClient, apolloProvider } from '../../../../tests/utils/mockGraphqlClient'
import store from '../../../store/index'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import config from '../../../lib/config'
import router from '../../../router'

Vue.use(Vuex)
sync(store, router)

const testCompound = (name: string, id: string) => ({
  name,
  imageURL: `/mol-images/HMDB/${id}.svg`,
  information: [{
    databaseId: id,
    database: 'HMDB-v4',
    url: `http://www.hmdb.ca/metabolites/${id}`,
  }],
})

const referenceAnnotation = {
  id: '2019-02-12_15h55m06s_HMDB-v4_2018-04-09_C10H11NO_plus_Na',
  ion: 'C10H11NO+Na+',
  fdrLevel: 0.1,
  dataset: { id: '2019-02-12_15h55m06s' },
  possibleCompounds: [testCompound('Tryptophanol', 'HMDB0003447')],
}
const relatedAnnotation = {
  ...referenceAnnotation,
  id: '2019-02-12_15h55m06s_HMDB-v4_2018-04-09_C10H13NO2_minus_H2O_plus_Na',
  ion: 'C10H13NO2-H2O+Na+',
  fdrLevel: 0.2,
  possibleCompounds: [
    testCompound('Maltoxazine', 'HMDB0030372'),
    testCompound('Salsolinol', 'HMDB0042012'),
  ],
}

describe('RelatedMolecules', () => {
  const propsData = { annotation: referenceAnnotation, databaseId: 22 }

  it('should match snapshot', async() => {
    config.features.isomers = true
    config.features.isobars = false
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([
          relatedAnnotation,
          referenceAnnotation,
        ]),
      }),
    })
    const wrapper = mount(RelatedMolecules, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.findAll('.ion-link').wrappers.map(el => el.text())).toEqual([
      expect.stringContaining('C10H11NO'),
      expect.stringContaining('C10H13NO2'),
    ])
    expect(wrapper.findAll('.compound').wrappers.map(el => el.text())).toEqual([
      expect.stringContaining('Tryptophanol'),
      expect.stringContaining('Maltoxazine'),
      expect.stringContaining('Salsolinol'),
    ])
  })

  it('should hide the isomer when the feature flag is off', async() => {
    config.features.isomers = false
    config.features.isobars = false
    initMockGraphqlClient({
      Query: () => ({
        allAnnotations: () => ([
          relatedAnnotation,
          referenceAnnotation,
        ]),
      }),
    })
    const wrapper = mount(RelatedMolecules, { store, router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper.find('.ion-link').exists()).toBe(false)
    expect(wrapper.findAll('.compound').wrappers.map(el => el.text())).toEqual([
      expect.stringContaining('Tryptophanol'),
    ])
  })
})
