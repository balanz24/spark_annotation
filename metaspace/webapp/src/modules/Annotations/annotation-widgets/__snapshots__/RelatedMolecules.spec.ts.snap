// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedMolecules should match snapshot 1`] = `
<div class="related-molecules" mock-v-loading="0">
  <div>
    <div class="el-divider el-divider--horizontal">
      <div class="el-divider__text is-center">
        <div class="ion-heading">
          <div class="ion-link">
            <div>
              <!----> <span class="ion-formula">[C<sub class="leading-none">10</sub>H<sub class="leading-none">11</sub>NO + Na]⁺</span></div>
            <div class="fdr-badge fdr-badge-10">
              10% FDR
            </div>
            <!---->
          </div>
          <!---->
        </div>
      </div>
    </div>
    <div class="compound-list">
      <div class="compound">
        <mock-el-popover placement="left" trigger="click">
          <div slot-key="default">
            <div class="compound-popover">
              <figure>
                <figcaption>
                  Tryptophanol
                  <br> <a href="http://www.hmdb.ca/metabolites/HMDB0003447" target="_blank">
                    View on HMDB-v4 website
                  </a></figcaption> <img src="/mol-images/HMDB/HMDB0003447.svg" class="compound-image">
              </figure>
            </div>
          </div>
          <div slot-key="reference"><img src="/mol-images/HMDB/HMDB0003447.svg" class="compound-thumbnail"></div>
        </mock-el-popover> <br> <span><a href="http://www.hmdb.ca/metabolites/HMDB0003447" target="_blank">
        Tryptophanol
      </a></span>
      </div>
      <!---->
    </div>
  </div>
  <div>
    <div class="el-divider el-divider--horizontal">
      <div class="el-divider__text is-center">
        <div class="ion-heading"><a href="/?ds=2019-02-12_15h55m06s&amp;mol=allAnnotations.0.sumFormula&amp;chem_mod=allAnnotations.0.chemMod&amp;nl=allAnnotations.0.neutralLoss&amp;add=allAnnotations.0.adduct&amp;fdr=0.2" class="ion-link">
            <div><span>Isomer:</span> <span class="ion-formula">[C<sub class="leading-none">10</sub>H<sub class="leading-none">13</sub>NO<sub class="leading-none">2</sub> + Na - H<sub class="leading-none">2</sub>O]⁺</span></div>
            <div class="fdr-badge fdr-badge-20">
              20% FDR
            </div>
            <!---->
          </a>
          <mock-el-popover trigger="hover" placement="top">
            <div slot-key="default">
              <div style="max-width: 500px;">
                <p>
                  The False Discovery Rate (FDR) for each annotation is calculated among all ions that share the same adduct.
                </p>
                <p>
                  It is possible for isomeric annotations to show different FDRs due to having different adducts.
                  In these cases, it is an indicator of certain adducts having a higher or lower probability of
                  incorrectly labelling unknown molecules.
                </p>
                <p>
                  The FDR should not be used to decide which isomeric molecule is more likely to be correct.
                </p>
              </div>
            </div>
            <div slot-key="reference"><i class="el-icon-question help-icon"></i></div>
          </mock-el-popover>
        </div>
      </div>
    </div>
    <div class="compound-list">
      <div class="compound">
        <mock-el-popover placement="left" trigger="click">
          <div slot-key="default">
            <div class="compound-popover">
              <figure>
                <figcaption>
                  Maltoxazine
                  <br> <a href="http://www.hmdb.ca/metabolites/HMDB0030372" target="_blank">
                    View on HMDB-v4 website
                  </a></figcaption> <img src="/mol-images/HMDB/HMDB0030372.svg" class="compound-image">
              </figure>
            </div>
          </div>
          <div slot-key="reference"><img src="/mol-images/HMDB/HMDB0030372.svg" class="compound-thumbnail"></div>
        </mock-el-popover> <br> <span><a href="http://www.hmdb.ca/metabolites/HMDB0030372" target="_blank">
        Maltoxazine
      </a></span>
      </div>
      <div class="compound">
        <mock-el-popover placement="left" trigger="click">
          <div slot-key="default">
            <div class="compound-popover">
              <figure>
                <figcaption>
                  Salsolinol
                  <br> <a href="http://www.hmdb.ca/metabolites/HMDB0042012" target="_blank">
                    View on HMDB-v4 website
                  </a></figcaption> <img src="/mol-images/HMDB/HMDB0042012.svg" class="compound-image">
              </figure>
            </div>
          </div>
          <div slot-key="reference"><img src="/mol-images/HMDB/HMDB0042012.svg" class="compound-thumbnail"></div>
        </mock-el-popover> <br> <span><a href="http://www.hmdb.ca/metabolites/HMDB0042012" target="_blank">
        Salsolinol
      </a></span>
      </div>
      <!---->
    </div>
  </div>
  <!---->
</div>
`;
